compiler/lib/effects.ml

```diff
@@ -355,35 +422,38 @@ let allocate_continuation ~st ~alloc_jump_closures ~split_closures pc x cont =
      closure to bind [x] if it is used in the loop body. In
      other cases, we can just pass the closure corresponding
      to the next block. *)
-  let pc', args = cont in
+  let direct_pc, args = direct_cont in
   if
     (match args with
     | [] -> true
     | [ x' ] -> Var.equal x x'
     | _ -> false)
     &&
-    match Hashtbl.find st.is_continuation pc' with
+    match Hashtbl.find st.is_continuation direct_pc with
     | `Param _ -> true
     | `Loop -> st.live_vars.(Var.idx x) = List.length args
-  then alloc_jump_closures, closure_of_pc ~st pc'
+  then alloc_jump_closures, closure_of_pc ~st direct_pc
   else
-    let body, branch = cps_branch ~st ~src:pc cont in
+    let body, branch = cps_branch ~st ~src:src_pc direct_cont in
     let inner_closures, outer_closures =
       (* For [Pushtrap], we need to separate the closures
          corresponding to the exception handler body (that may make
          use of [x]) from the other closures that may be used outside
          of the exception handler. *)
       if not split_closures
       then alloc_jump_closures, []
-      else if is_merge_node st.cfg pc'
+      else if is_merge_node st.cfg direct_pc
       then [], alloc_jump_closures
       else
-        List.partition
-          ~f:(fun i ->
-            match i with
-            | Let (_, Closure (_, (pc'', []))) -> dominates st.cfg st.idom pc' pc''
```

Comment:
We can revive this code, using this condition instead:
```ocaml
            | Let (_, Closure (_, (pc'', []))) -> pc'' = mk_cps_pc_of_direct ~st direct_pc
```
Indeed, we have `Hashtbl.find idom pc'' = src_pc` (by construction of `closures_of_alloc_site`). And `direct_pc` cannot dominate `src_pc`  since it comes afterwards. So, in the code below, the second case `dominates g idom direct_pc (Hashtbl.find idom pc'')` will always be false.
```ocaml
let rec dominates g idom pc pc' =
  pc = pc'
  || Hashtbl.find g.block_order pc < Hashtbl.find g.block_order pc'
     && dominates g idom pc (Hashtbl.find idom pc')
```

Reply:
Should I still apply `do_alloc_jump_closures` to both resulting lists?

Reply:
Should I still apply `do_alloc_jump_closures` to both resulting lists?

Reply:
No. The point is that you can work on `alloc_jump_closures`.